### PR TITLE
Fix locale for aperture filter

### LINKED ISSUE #16031: The collection selection does not work correctly
### Describe the bug

I am solving the problem of how many times I used an aperture in the range of 2.8 to 4 with a Tamron 24-70 lens on a Nikon D750 while rating the image with 3 stars and a red mark at the same time. As long as I don't enter the aperture selection, everything filters fine. As soon as I try to add aperture selection nothing is found.
Example:
collection: Nikon-D750
rating: 3 stars
color mark: red
lens: Tamron 24-70mm ..
aperture: 2.8 -- it doesn't pick anything here

Am I doing something wrong or where is the problem?
Thanks for the direction if I'm doing it wrong.

### Steps to reproduce

1. collection: Nikon-D750
2. rating: 3 stars
3. color mark: red
4. lens: Tamron 24-70mm ..
5. aperture: 2.8 -- it doesn't pick anything here

### Expected behavior

_No response_

### Logfile | Screenshot | Screencast

_No response_

### Commit

_No response_

### Where did you obtain darktable from?

downloaded from www.darktable.org

### darktable version

4.6.0

### What OS are you using?

Linux

### What is the version of your OS?

openSUSE Leap 15.5

### Describe your system?

_No response_

### Are you using OpenCL GPU in darktable?

Yes

### If yes, what is the GPU card and driver?

_No response_

### Please provide additional context if applicable. You can attach files too, but might need to rename to .txt or .zip

_No response_

### PR DESCRIPTION
**Pull request by zisoft**
fixes #16031 

The locale needs to be switched for the aperture values, similar to the exposure filter.

### CODE REVIEW

File: src/libs/filters/aperture.c [hunk 1]
```diff
@@ -67,7 +67,19 @@ static gboolean _aperture_update(dt_lib_filtering_rule_t *rule)
 
 static gchar *_aperture_print_func(const double value, const gboolean detailled)
 {
-  return g_strdup_printf("%s%.1lf", detailled ? "f/" : "", value);
+  if (detailled)
+  {
+    return g_strdup_printf("%s%.1lf", detailled ? "f/" : "", value);
```
Comment (TurboGit) [inline src/libs/filters/aperture.c:72]: You can then probably remove `detailled` check here. But I'm wondering why we don't need a local switch in detailled mode?

Comment (zisoft) [reply src/libs/filters/aperture.c:72]: The `detailled` is used when the list is filled with all available aperture values (right click in the text field).

Bildschirmfoto 2024-01-06 um 12 22 06

Comment (zisoft) [reply src/libs/filters/aperture.c:72]: In this case it is only the textural representation, no values for computing.

File: src/libs/filters/aperture.c [hunk 2]
```diff
@@ -67,7 +67,19 @@ static gboolean _aperture_update(dt_lib_filtering_rule_t *rule)
 
 static gchar *_aperture_print_func(const double value, const gboolean detailled)
 {
-  return g_strdup_printf("%s%.1lf", detailled ? "f/" : "", value);
+  if (detailled)
```
Comment (TurboGit) [inline src/libs/filters/aperture.c:70]: no space after if

File: src/libs/filters/aperture.c [hunk 3]
```diff
@@ -67,7 +67,19 @@ static gboolean _aperture_update(dt_lib_filtering_rule_t *rule)
 
 static gchar *_aperture_print_func(const double value, const gboolean detailled)
 {
-  return g_strdup_printf("%s%.1lf", detailled ? "f/" : "", value);
+  if(detailled)
+  {
+    return g_strdup_printf("%s%.1lf", detailled ? "f/" : "", value);
```
Comment (TurboGit) [inline src/libs/filters/aperture.c:72]: But we still can remove `detailled` here as we are already in the conditional `if(detailled)`.

File: src/libs/filters/aperture.c [hunk 4]
```diff
@@ -67,7 +67,19 @@ static gboolean _aperture_update(dt_lib_filtering_rule_t *rule)
...
+  }
+  else
+  {
+    gchar *locale = g_strdup(setlocale(LC_ALL, NULL));
+    setlocale(LC_NUMERIC, "C");
```
Comment (parafin) [inline src/libs/filters/aperture.c:77]: setlocale is not thread-safe, this is not a correct way to solve this.

File: src/libs/filters/aperture.c (hunk 3 above)
Comment (zisoft) [reply src/libs/filters/aperture.c]: ahh, now I see you point :-)
Didn't touch that original line.

File: src/libs/filters/aperture.c (hunk 4 above)
Comment (TurboGit) [reply src/libs/filters/aperture.c:77]: All this is happening from the main Gtk GUI thread as far as I understand, we should be safe, no?

Comment (parafin) [reply src/libs/filters/aperture.c:77]: In C++ one can use stringstream and imbue() to set locale just for that stream. I’m not sure if anything simular exists in C. My suggestion is to create a helper function (to be used from C code) to format floating point number with dot separator and implement it in C++.

Comment (parafin) [reply src/libs/filters/aperture.c:77]: > All this is happening from the main Gtk GUI thread as far as I understand, we should be safe, no?

Are you sure no other threads use functions like printf? Even threads created by libraries?

No, it’s not safe.

Comment (TurboGit) [reply src/libs/filters/aperture.c:77]: Right, just after my message it occurs to me that some export could be running in background and so we will change the local also for them.

Comment (zisoft) [reply src/libs/filters/aperture.c:77]: Ok, then we should use `duplocale` and `newlocale`.
This is needed for the exposure filter as well.

Comment (parafin) [reply src/libs/filters/aperture.c:77]: Right, uselocale is a C-way to solve this.
And FYI this is a kind of issues created by setlocale: https://github.com/darktable-org/darktable/issues/13637#issuecomment-1431006172

Comment (parafin) [reply src/libs/filters/aperture.c:77]: Also be advised that on Windows it may require different code, see linked lensfun issue.

Comment (parafin) [reply src/libs/filters/aperture.c:77]: https://github.com/lensfun/lensfun/pull/1987/files

Comment (zisoft) [reply src/libs/filters/aperture.c]: I have now created a function in `common/utility.c`.
Tested on Windows as well.

BTW: There are several other calls of `setlocale` in the code base which should be adapted.

Comment (zisoft) [reply src/libs/filters/aperture.c]: Fixed the other filters as well: exposure, iso, ratio

File: src/common/utility.c [hunk 5]
```diff
@@ -97,6 +97,28 @@ guint dt_util_str_occurence(const gchar *haystack, const gchar *needle)
   return o;
 }
 
+gchar *dt_util_float_to_str(const gchar *format, const double value)
+{
+#if defined(WIN32)
+  _configthreadlocale(_ENABLE_PER_THREAD_LOCALE);
+  setlocale (LC_NUMERIC, "C");
+#else
+  locale_t locale = uselocale((locale_t) 0);
```
Comment (TurboGit) [inline src/common/utility.c:106]: Why not simply:

```
  locale_t nlocale = newlocale(LC_NUMERIC_MASK, "C", (locale_t) 0);
  locale_t locale = uselocale(nlocale);
```

Comment (zisoft) [reply src/common/utility.c]: Indeed, it saves a function call.

It's hard to get you satisfied 😉
Your review work is incredible good!

Comment (TurboGit) [reply src/common/utility.c]: Thanks for the nice comment @zisoft, yet I'm missing quite some issues :) It is hard to spot everything.

**Review by TurboGit — APPROVED** [review]
Looks good to me now. Thanks!

**Comment by TurboGit** [comment]
Squashed and merged in master & 4.6.x branch.

**Comment by zisoft** [comment]
For the release notes:
Fix collection aperture filter not working on some locales.

**Pull request merged**